openvdb_points/tools/AttributeGroup.h

```diff
@@ -0,0 +1,226 @@
+{
+public:
+    /// Default constructor, always constructs a uniform attribute.
+    explicit GroupAttributeArray(   size_t n = 1,
+                                    const ValueType& uniformValue = zeroVal<ValueType>());
+    /// Deep copy constructor (optionally decompress during copy).
+    GroupAttributeArray(const TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >& array,
+                        const bool decompress = false);
+
+    /// Cast an AttributeArray to GroupAttributeArray
+    static GroupAttributeArray& cast(AttributeArray& attributeArray);
+
+    /// Cast an AttributeArray to GroupAttributeArray
+    static const GroupAttributeArray& cast(const AttributeArray& attributeArray);
+
+    /// Return @c true if the AttributeArray provided is a group
+    static bool isGroup(const AttributeArray& attributeArray);
+
+    /// @brief Specify whether this attribute is for tracking group membership
+    /// @note  Attributes are not group attributes by default.
+    void setGroup(bool state);
+    /// Return @c true if this attribute is for tracking groups
+    bool isGroup() const { return bool(mFlags & GROUP); }
+
+}; // class GroupAttributeArray
+
+
+inline GroupAttributeArray&
+GroupAttributeArray::cast(AttributeArray& attributeArray)
+{
+    if (!attributeArray.isType<GroupAttributeArray>()) {
+        OPENVDB_THROW(TypeError, "Invalid Attribute Type");
+    }
+    return static_cast<GroupAttributeArray&>(attributeArray);
+}
+
+
+inline const GroupAttributeArray&
+GroupAttributeArray::cast(const AttributeArray& attributeArray)
+{
+    if (!attributeArray.isType<GroupAttributeArray>()) {
+        OPENVDB_THROW(TypeError, "Invalid Attribute Type");
+    }
+    return static_cast<const GroupAttributeArray&>(attributeArray);
+}
+
+
+inline bool
+GroupAttributeArray::isGroup(const AttributeArray& attributeArray)
+{
+    if (!attributeArray.isType<GroupAttributeArray>())  return false;
+
+    return GroupAttributeArray::cast(attributeArray).isGroup();
+}
+
+
+////////////////////////////////////////
+
+
+class GroupHandle
+{
+public:
+    // Dummy class that distinguishes an offset from a bitmask on construction
+    struct BitMask { };
+
+    typedef std::pair<size_t, uint8_t> GroupIndex;
+
+    GroupHandle(const GroupAttributeArray& array,
+                const GroupType& offset)
+        : mArray(array)
+        , mBitMask(GroupType(1) << offset) { assert(mArray.isGroup()); }
+
+    GroupHandle(const GroupAttributeArray& array,
+                const GroupType& bitMask,
+                BitMask)
+        : mArray(array)
+        , mBitMask(bitMask) { assert(mArray.isGroup()); }
+
+    size_t size() const { return mArray.size(); }
+
+    bool get(Index n) const;
+
+private:
+    const GroupAttributeArray& mArray;
+    const GroupType mBitMask;
+}; // class GroupHandle
+
+
+////////////////////////////////////////
+
+
+class GroupWriteHandle
+{
```

Comment:
Re-written this to use a similar implementation to the AttributeHandles, definitely better, thanks.